schema/query/predicate.go

```diff
@@ -303,97 +333,130 @@ func (e NotExist) Validate(validator schema.Validator) error {
 }
 
 // GreaterOrEqual matches values that are greater than or equal to a specified value.
 type GreaterOrEqual struct {
 	Field string
-	Value float64
+	Value Value
 }
 
 // Match implements Expression interface
 func (e GreaterOrEqual) Match(payload map[string]interface{}) bool {
-	n, ok := isNumber(getField(payload, e.Field))
-	return ok && (n >= e.Value)
+	// n, ok := isNumber(getField(payload, e.Field))
+	// return ok && (n >= e.Value)
+	return true
 }
 
 // Validate implements Expression interface.
 func (e GreaterOrEqual) Validate(validator schema.Validator) error {
-	return validateNumericValue(e.Field, e.Value, opGreaterOrEqual, validator)
+	return validateValue(e.Field, e.Value, schema.OpGreaterOrEqual, validator)
+}
+
+// ValidateOperator TODO
+func (e GreaterOrEqual) ValidateOperator(validator schema.Validator) (Expression, error) {
+	newValue, err := validateValueOperator(e.Field, e.Value, schema.OpGreaterOrEqual, validator)
+	return GreaterOrEqual{Field: e.Field, Value: newValue}, err
 }
 
 // String implements Expression interface.
 func (e GreaterOrEqual) String() string {
-	return quoteField(e.Field) + ": {" + opGreaterOrEqual + ": " + valueString(e.Value) + "}"
+	return quoteField(e.Field) + ": {" + schema.OpGreaterOrEqual + ": " + valueString(e.Value) + "}"
 }
 
 // LowerThan matches values that are less than a specified value.
 type LowerThan struct {
 	Field string
-	Value float64
+	Value Value
 }
 
 // Match implements Expression interface.
 func (e LowerThan) Match(payload map[string]interface{}) bool {
-	n, ok := isNumber(getField(payload, e.Field))
-	return ok && (n < e.Value)
+	// n, ok := isNumber(getField(payload, e.Field))
+	// return ok && (n < e.Value)
+	return true
 }
 
 // Validate implements Expression interface.
 func (e LowerThan) Validate(validator schema.Validator) error {
-	return validateNumericValue(e.Field, e.Value, opLowerThan, validator)
+	return validateValue(e.Field, e.Value, schema.OpLowerThan, validator)
+}
+
+// ValidateOperator TODO
+func (e LowerThan) ValidateOperator(validator schema.Validator) (Expression, error) {
+	newValue, err := validateValueOperator(e.Field, e.Value, schema.OpLowerThan, validator)
+	return LowerThan{Field: e.Field, Value: newValue}, err
 }
 
 // String implements Expression interface.
 func (e LowerThan) String() string {
-	return quoteField(e.Field) + ": {" + opLowerThan + ": " + valueString(e.Value) + "}"
+	return quoteField(e.Field) + ": {" + schema.OpLowerThan + ": " + valueString(e.Value) + "}"
 }
 
 // LowerOrEqual matches values that are less than or equal to a specified value.
 type LowerOrEqual struct {
 	Field string
-	Value float64
+	Value Value
 }
 
 // Match implements Expression interface.
 func (e LowerOrEqual) Match(payload map[string]interface{}) bool {
-	n, ok := isNumber(getField(payload, e.Field))
-	return ok && (n <= e.Value)
+	// n, ok := isNumber(getField(payload, e.Field))
+	// return ok && (n <= e.Value)
```

Comment:
Yes, I guess there is a lot of work for `rest-layer-mem` to support all MongoDB functionality and pecularities. For now my intention was to just hard-code support for int/float/Time into `Match` function. But I can try your proposal here.

Reply:
For the `a == b` and `a != b` case I guess we have more generic implementation using `reflect.DeepEqual`. Shouldn't we keep it. Also if a type implements `Comparable` then it would support `$gt $gte $lt $lte` implicitly so there is no need to restate that in `ValidateOprator` method for each type.

Reply:
@smyrman Any comment on the above?

Reply:
Hi @Dragomir-Ivanov . That may be a good approach I suppose -- it doens't solve the case of regex, but perhaps that should be solved separately?

On the schema level, would it then be that a FieldValidator implementation that is to support `$lt[e]`/`$gt[e]` implements an interface ala:

```go
type ComparableField struct {
    /// ComparableValue returns a comparable and true, or nil and false.
    ComparableValue(value interface{}) (Comparable, bool)
}
```

Technically, it doesn't need the "ok" `bool` return, it could just return `nil` when it's not possible to convert to the target value, but keeping the second parameter kind of makes the intent more clear, and that it can return nil.

Same reasoning goes for the `AsComparable`. I don't know what's the best design, but as long as the behaviour is documented, I would accept either.

And maybe `string` types that supports Regex search could implement (different PR):

```go
type RegexField struct {
    CompareableToRegex() bool
}
```
```go
type Comparable interface {
	// AsComparable converts other to a comparable type if possible.
	AsComparable(other interface{}) (Comparable, bool)

	// Less returns true if other is comparable and less than this value.
	Less(other interface{}) bool
}
```

Reply:
It doesn't change the need to pass a schema.Validator to the `query.Predicate` in order to get the comparable value, but it may be a bit more clear in the `schema` pacakge, and we won't need to move or expose all the operator value definitions.

Reply:
Closing this conversation, as there is now a better proposal for the interface.